File: server/src/main/com/broll/gainea/server/core/fractions/impl/KnightsFraction.java
package com.broll.gainea.server.core.fractions.impl;

import com.broll.gainea.server.core.battle.BattleContext;
import com.broll.gainea.server.core.battle.FightingPower;
import com.broll.gainea.server.core.fractions.Fraction;
import com.broll.gainea.server.core.fractions.FractionDescription;
import com.broll.gainea.server.core.fractions.FractionType;
import com.broll.gainea.server.core.map.AreaType;
import com.broll.gainea.server.core.objects.Commander;
import com.broll.gainea.server.core.objects.Soldier;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.utils.LocationUtils;

import java.util.List;

public class KnightsFraction extends Fraction {

    public KnightsFraction() {
        super(FractionType.KNIGHTS);
    }

    @Override
    protected FractionDescription description() {
        FractionDescription desc = new FractionDescription("");
        desc.plus("In Unterzahl +1 Zahl");
        desc.plus("Im 1v1 +2 Zahl");
        desc.contra("Auf Sümpfen und Seen -1 Zahl");
        return desc;
    }

    @Override
    public FightingPower calcFightingPower(Soldier soldier, BattleContext context) {
        FightingPower power = super.calcFightingPower(soldier, context);
        List<Unit> army = context.getFightingArmy(soldier);
        List<Unit> opponents = context.getOpposingFightingArmy(soldier);
        if (army.size() < opponents.size()) {
            // smaller army +1 Z
            power.changeNumberPlus(1);
        } else if (army.size() == 1 && opponents.size() == 1) {
            //1v1  +2 Z
            power.changeNumberPlus(2);
        }
        if (LocationUtils.isAreaType(context.getLocation(), AreaType.BOG, AreaType.LAKE)) {
            power.changeNumberPlus(-1);
        }
        return power;
    }

    @Override
    public Soldier createSoldier() {
        Soldier soldier = new Soldier(owner);
        soldier.setStats(SOLDIER_POWER, SOLDIER_HEALTH);
        soldier.setName("Kreuzritter");
        soldier.setIcon(11);
        return soldier;
    }


    @Override
    public Commander createCommander() {
        Commander commander = new Commander(owner);
        commander.setStats(COMMANDER_POWER, COMMANDER_HEALTH);
        commander.setName("Kreuzritterchampion");
        commander.setIcon(7);
        return commander;
    }

}

File: server/src/main/com/broll/gainea/server/core/fractions/impl/KnightsFraction.kt
package com.broll.gainea.server.core.fractions.impl

import com.broll.gainea.server.core.battle.BattleContext
import com.broll.gainea.server.core.battle.FightingPower
import com.broll.gainea.server.core.fractions.Fraction
import com.broll.gainea.server.core.fractions.FractionDescription
import com.broll.gainea.server.core.fractions.FractionType
import com.broll.gainea.server.core.fractions.UnitDescription
import com.broll.gainea.server.core.map.AreaType
import com.broll.gainea.server.core.objects.Soldier
import com.broll.gainea.server.core.utils.isAreaType

class KnightsFraction : Fraction(FractionType.KNIGHTS) {

    override val description: FractionDescription = FractionDescription(
        "",
        soldier = UnitDescription(name = "Kreuzritter", icon = 11),
        commander = UnitDescription(
            name = "Kreuzritterchampion",
            icon = 7,
            power = 3,
            health = 3
        ),
    ).apply {
        plus("Im 1v1 +1 Zahl, als Angreifer +2 Zahl")
        contra("Auf Sümpfen -1 Zahl")
    }

    override fun calcFightingPower(soldier: Soldier, context: BattleContext): FightingPower {
        val power = super.calcFightingPower(soldier, context)
        val army = context.getFightingArmy(soldier)
        val opponents = context.getOpposingFightingArmy(soldier)
        if (army.size == 1 && opponents.size == 1) {
            //1v1  +2 Z (atk) / +1 Z (def)
            val plus = if (context.isAttacking(army.first())) 2 else 1
            power.changeNumberPlus(plus)
        }
        if (context.location.isAreaType(AreaType.BOG)) {
            power.changeNumberPlus(-1)
        }
        return power
    }

}
